lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -174,116 +233,150 @@ public static Domain getDomain(
+    @VisibleForTesting
+    public Domain getDomain(Type type, long rowCount, ColumnIndex columnIndex, ParquetDataSourceId id, RichColumnDescriptor descriptor)
             throws ParquetCorruptionException
     {
-        if (statistics instanceof LongStatistics) {
-            LongStatistics longStatistics = (LongStatistics) statistics;
-            if (longStatistics.genericGetMin() > longStatistics.genericGetMax()) {
-                throw corruptionException(column, id, longStatistics);
-            }
-            return Optional.of(new ParquetIntegerStatistics(longStatistics.genericGetMin(), longStatistics.genericGetMax()));
+        if (columnIndex == null) {
+            return Domain.all(type);
         }
 
-        if (statistics instanceof IntStatistics) {
-            IntStatistics intStatistics = (IntStatistics) statistics;
-            if (intStatistics.genericGetMin() > intStatistics.genericGetMax()) {
-                throw corruptionException(column, id, intStatistics);
+        String columnName = descriptor.getPrimitiveType().getName();
+
+        if (isCorruptedColumnIndex(columnIndex)) {
+            throw corruptionException(columnName, id, columnIndex, null);
+        }
+
+        if (isEmptyColumnIndex(columnIndex)) {
+            return Domain.all(type);
+        }
+
+        long totalNullCount = columnIndex.getNullCounts().stream()
+                .mapToLong(value -> value)
+                .sum();
+        boolean hasNullValue = totalNullCount > 0;
+
+        if (hasNullValue && totalNullCount == rowCount) {
+            return Domain.onlyNull(type);
+        }
+
+        try {
+            Domain domain = Domain.none(type);
+            int pageCount = columnIndex.getMinValues().size();
+            ColumnIndexValueConverter converter = new ColumnIndexValueConverter();
+            Function<ByteBuffer, Object> converterFunction = converter.getConverter(descriptor.getPrimitiveType());
+            for (int i = 0; i < pageCount; i++) {
+                Object min = converterFunction.apply(columnIndex.getMinValues().get(i));
+                Object max = converterFunction.apply(columnIndex.getMaxValues().get(i));
+                domain = domain.union(getDomain(type, min, max, hasNullValue, timeZone));
             }
-            return Optional.of(new ParquetIntegerStatistics((long) intStatistics.getMin(), (long) intStatistics.getMax()));
+
+            return domain;
+        }
+        catch (Exception e) {
+            throw corruptionException(columnName, id, columnIndex, e);
+        }
+    }
+
+    private static ParquetLongStatistics toParquetLongStatistics(Object min, Object max)
+    {
+        if (min instanceof Number) {
+            Number minValue = (Number) min;
+            Number maxValue = (Number) max;
+            return new ParquetLongStatistics(minValue.longValue(), maxValue.longValue());
+        }
```

Comment:
It'd be better to check for the possible concrete types explicitly. `longValue` on `Number` can cause loss of precision for some data types such as Double. If Double is not allowed or expected, the method shouldn't even try to handle it.

Reply:
Added check for Double or Float and throw now instead